### PR TITLE
Better error handling for v2

### LINKED ISSUE #2: Mocka needs better error handling
## Context / Problem
Many methods with `mocka` have the potential to return an error. Once the initial setup error has been resolved `mocka` will not throw runtime errors.

As a user, being forced to check for errors as bloat to the code base.

> Note: Many error cases are hidden in `mocka` and that may result in `mocka` not acting as you might expect.

## Proposal 
- [x] Move public interfaces from `mocka` reducing abstractions
- [x] Remove all returned error and have `mocka` internally fail the test
- [x] Update `mocka.Function()` to take in a new interface for `testing.T` to assume in order to fail the test.

### PR DESCRIPTION
**Pull request by cmarfia**
# CHANGELOG

## Changed
- `mocka.Function` now takes in a `TestReporter` to fail tests internally inside of mocka
- `mocka.CreateSandbox` now takes in a `TestReporter` to fail tests internally inside of mocka
- `Sandbox.StubFunction` was renamed to `Sandbox.Function` for consistency
- `mocka.Function` and `mocka.CreateSandbox` will call `log.Fatal` if a `TestReporter` is not provided
- Exposed internal types to remove the need to unnecessary interfaces
- `Stub.Return` no longers returns an error, but instead fails the test internally
- `Stub.GetCall` no longer panics, but instead fails the test internally
- Updated documentation to reflect public API changes
- Module is now updated to have v2 suffix
- All internal imports reference mocka/v2

## Removed
- `argumentValidationError`
- `outParameterValidationError`
- All public interfaces that were only used for defining the API

### CODE REVIEW

File: customArguments.go [hunk 1]
```diff
@@ -1,38 +1,29 @@
...
+func newCustomArguments(stub *Stub, arguments []interface{}) *CustomArguments {
 	functionType := stub.toType()
 	if isArgumentLengthValid(functionType, arguments) {
-		return &customArguments{
-			argValidationError: &argumentValidationError{
-				fnType:   functionType,
-				provided: arguments,
-			},
-		}
+		stub.testReporter.Errorf("%v", &argumentValidationError{fnType: functionType, provided: arguments})
```
Comment (cmarfia) [inline customArguments.go]: Should this be refactored to no longer be an error?

File: customArguments.go [hunk 2]
```diff
@@ -111,81 +102,71 @@ func getMatcher(value interface{}, valueType reflect.Type) (match.SupportedKinds
...
 // Return sets the return values for this set of custom arguments
-func (ca *customArguments) Return(returnValues ...interface{}) error {
-	if ca.argValidationError != nil {
-		return ca.argValidationError
-	}
-
-	if ca.stub == nil {
-		return errors.New("mocka: stub does not exist")
+func (ca *CustomArguments) Return(returnValues ...interface{}) {
+	if !validateOutParameters(ca.stub.toType(), returnValues) {
```
Comment (cmarfia) [inline customArguments.go:110]: Should this return an error instead of a boolean?

File: customArguments.go [hunk 3]
```diff
@@ -111,81 +102,71 @@ func getMatcher(value interface{}, valueType reflect.Type) (match.SupportedKinds
...
-
+func (ca *CustomArguments) OnCall(callIndex int) *OnCall {
 	for _, o := range ca.onCalls {
 		if o.index == callIndex {
 			return o
 		}
 	}
 
-	o := &onCall{index: callIndex, stub: ca.stub}
+	ca.stub.lock.Lock()
```
Comment (cmarfia) [inline customArguments.go]: These locks should be moved back above the for loop.

File: customArguments_test.go [hunk 4]
```diff
@@ -9,99 +9,69 @@ import (
...
-
-			Expect(ca).ToNot(BeNil())
-			Expect(*ca).To(Equal(customArguments{
-				argValidationError: &argumentValidationError{
-					provided: []interface{}{},
-				},
-			}))
-		})
-
 		It("returns with validation error when provided arguments != stubbed function arguments length", func() {
```
Comment (cmarfia) [inline customArguments_test.go]: Test name needs to be updated.

File: customArguments_test.go [hunk 5]
```diff
@@ -9,99 +9,69 @@ import (
...
-					provided: []interface{}{},
-				},
+			_ = newCustomArguments(stub, []interface{}{})
+
+			Expect(failTestReporter.messages).To(Equal([]string{
+				"mocka: expected arguments of type (string, int), but received ()",
 			}))
 		})
 
 		It("returns with validation error if the provided matcher is not supported for the argument kind", func() {
```
Comment (cmarfia) [inline customArguments_test.go]: Test name needs to be updated.

File: customArguments_test.go [hunk 6]
```diff
@@ -9,99 +9,69 @@ import (
...
-				argValidationError: &argumentValidationError{
-					fnType:   mockFn.toType(),
-					provided: []interface{}{anything, lengthOf10},
-				},
+			Expect(failTestReporter.messages).To(Equal([]string{
+				"mocka: expected arguments of type (string, int), but received (*anything, *lengthOf)",
 			}))
 		})
 
 		It("returns with validation error if the provided argument is not of the correct type", func() {
```
Comment (cmarfia) [inline customArguments_test.go]: Test name needs to be updated.

File: customArguments_test.go [hunk 7]
```diff
@@ -110,170 +80,137 @@ var _ = Describe("customArguments", func() {
 			fn := func(msg string) error {
```
Comment (cmarfia) [inline customArguments_test.go:80]: Test name needs to be updated.

File: customArguments_test.go [hunk 8]
```diff
@@ -110,170 +80,137 @@ var _ = Describe("customArguments", func() {
...
-					stub: mockFn,
+				Expect(*ca).To(Equal(CustomArguments{
+					stub: stub,
 					argMatchers: []match.SupportedKindsMatcher{
 						match.Exactly("hi"),
 						match.SliceOf(match.Nil(), match.Exactly("A"), match.Anything())},
 				}))
 			})
 
 			It("returns an argument validation error if matcher does not suppor the variadic type", func() {
```
Comment (cmarfia) [inline customArguments_test.go]: Test name needs to be updated.

File: customArguments_test.go [hunk 9]
```diff
@@ -110,170 +80,137 @@ var _ = Describe("customArguments", func() {
...
-				},
-			}
-
-			err := ca.Return(42, "nil")
-
-			Expect(err).Should(HaveOccurred())
-			Expect(err.Error()).To(Equal("mocka: expected arguments of type (string, int), but received (int, int)"))
-		})
-
 		It("returns an error if one or more of the return values are not valid", func() {
```
Comment (cmarfia) [inline customArguments_test.go:166]: Test name needs to be updated.

File: mocka.go [hunk 10]
```diff
@@ -32,20 +32,37 @@
...
-func CreateSandbox() Sandbox {
-	return &sandbox{}
+func CreateSandbox(testReporter TestReporter) *Sandbox {
+	return &Sandbox{testReporter: ensureTestReporter(testReporter)}
+}
+
+// ensureTestReporter returns the existing test reporter or a new logger to Stderr
+func ensureTestReporter(testReporter TestReporter) TestReporter {
+	if testReporter == nil {
+		log.Fatal("mocka: TODO")
```
Comment (cmarfia) [inline mocka.go]: TODO

File: stub.go [hunk 11]
```diff
@@ -1,85 +1,56 @@
...
 	execFunc      func([]interface{})
 }
 
-// newMockFunction creates a new mock function and overrides the implementation of the original function.
-func newMockFunction(originalFuncPtr interface{}, returnValues []interface{}) (*mockFunction, error) {
+// newStub creates a stub function and overrides the implementation of the original function.
+func newStub(testReporter TestReporter, originalFuncPtr interface{}, returnValues []interface{}) *Stub {
 	if originalFuncPtr == nil {
-		return nil, errors.New("mocka: expected the first argument to be a pointer to a function, but received a nil")
+		testReporter.Errorf("mocka: expected the first argument to be a pointer to a function, but received a nil")
```
Comment (cmarfia) [inline stub.go]: Need to update error message to be second param.

The same for the following error messages

File: stub.go [hunk 12]
```diff
@@ -286,15 +258,16 @@ func (mf *mockFunction) GetCalls() []Call {
...
-	mf.lock.RLock()
-	defer mf.lock.RUnlock()
+func (stub *Stub) GetCall(callIndex int) Call {
+	stub.lock.RLock()
+	defer stub.lock.RUnlock()
 
-	if callIndex < 0 || callIndex >= mf.CallCount() {
-		panic(fmt.Errorf("mocka: attempted to get CallMetaData for call %v, when the function has only been called %v times", callIndex, len(mf.calls)))
+	if callIndex < 0 || callIndex >= stub.CallCount() {
+		stub.testReporter.Errorf("mocka: attempted to get CallMetaData for call %v, when the function has only been called %v times", callIndex, len(stub.calls))
```
Comment (cmarfia) [inline stub.go]: CallMetaData is not what it is called anymore

**Comment by cmarfia** [comment]
### TODO

- [x] Doc comment updates
- [x] Possible move of examples to their specific packages
- [x] README documentation updates

**Comment by cmarfia** [comment]
I Approve!

File: MATCH.md [hunk 13]
```diff
@@ -1,6 +1,10 @@
-# mocka/match <!-- omit in toc -->
+# mocka/match
 
-The package mocka/match provides a powerful suite of matchers that can be used in conjunction with the `WithArgs` function for a mocka Stub.
+```go
+import "github.com/MonsantoCo/mocka/match"
```
Comment (poost-frey) [inline MATCH.md]:
```suggestion
import "github.com/MonsantoCo/mocka/v2/match"
```

File: README.md [hunk 14]
```diff
@@ -1,383 +1,395 @@
-# mocka <!-- omit in toc -->
+# mocka [![Build Status][build-badge]][build-ci] [![gopherbadger-tag-do-not-edit][coverage-badge]][coverage] [![GoDoc][godoc-badge]][godoc]
 
-The package mocka provides a simple mocking and stubbing library to assist with writing unit tests in go.
+```go
+import "github.com/MonsantoCo/mocka"
```
Comment (poost-frey) [inline README.md]:
```suggestion
import "github.com/MonsantoCo/mocka/v2"
```

File: README.md [hunk 15]
```diff
@@ -1,383 +1,395 @@
...
-stub, err := mocka.Function(&fn, 20)
-if err != nil {
-    log.Fatal(err.Error())
-}
-defer stub.Restore()
+import (
+    "testing"
 
-fmt.Println(fn("123"))
+    "github.com/MonsantoCo/mocka"
```
Comment (poost-frey) [inline README.md]:
```suggestion
    "github.com/MonsantoCo/mocka/v2"
```
I won't mark any other occurrences of this, but its probably worth while to update them all together.

File: customArguments_test.go [hunk 16]
```diff
@@ -285,146 +222,125 @@ var _ = Describe("customArguments", func() {
...
 			}
 
 			result := ca.OnSecondCall()
 
-			o, ok := result.(*onCall)
-
 			Expect(ca.onCalls).To(HaveLen(3))
-			Expect(ok).To(BeTrue())
-			Expect(*o).To(Equal(onCall{stub: mockFn, index: 1}))
+			Expect(result).To(Equal(&OnCall{stub: stub, index: 1}))
```
Comment (poost-frey) [inline customArguments_test.go]: is the intention for the values to match and not the addresses? I think ginkgo is using the underlying values anyway. Might be more readable to do `Expect(*result).To(Equal(OnCall{stub: stub, index: 1}))`

File: README.md (hunk 15 above)
Comment (cmarfia) [reply README.md]: Good catch on these, I never updated to v2 in the go mod either 🤦

**Review by poost-frey — APPROVED** [review]

**Pull request merged**